File: dijkstra.c
/*
 * Dijkstra's algorithm.   
 */
#include <stdio.h>

int output[50], ind = 0;

const long PInfinit = 888;

int n;//in this variable we hold the number of nodes.

int road[50][50], R[50], F[50], S[50];

FILE *fin, *fout; //FILE I/O handler for files

int ender;

//prototypes
void read(const char *);
void writeToFile(const char *);
void dijkstra();
void solution(int);
void get(int);

int main() {

    printf("Arrival Node=");
    scanf("%d",&ender);
     
    read("road.in");
    dijkstra();
    solution( ender );

    writeToFile("dijkstra.out");

    return(0);
}

void read(const char* filename) {

     int i,j,x,y;
     float c;

     fin = fopen(filename, "r");

     fscanf(fin, "%d", &n);

     for(i=1;i<=n;i++) {

         for(j=1;j<=n;j++) {

             if(i == j) {

                  road[i][j] = 0; 

             } else {

                 road[i][j] = PInfinit;  
             }
         }
     }
     
     while(!feof(fin)) { 
            fscanf(fin, "%d %d %f", &x, &y, &c); 
            road[x][y] = c;
     }

     for(i=1;i<=n;i++) {

         for(j=1;j<=n;j++) {

             printf("%5d ", road[i][j]);
         }
             printf("\n");
     }    
}

void dijkstra() {
     int i,j,k,
           start,
                min,
                    pos = 0;

     start = 1;
      
     for(i=1;i<=n;i++) {
         R[i] = road[start][i];

         if(i!=start) 
            if(R[i] < PInfinit) F[i] = start;  
     }      
  
     S[start] = 1; 
   
     //execute n-1 times
     for(i=1;i<=n-1;i++) {
        
         min = PInfinit;

         for(j=1;j<=n;j++) {
            if(S[j] == 0)
             if(R[j]<min){
                min = R[j];
                pos = j;
             } 
         }        

         S[ pos ] = 1;

         for(j=1;j<=n;j++) {

             if(S[j] == 0) 

                if(R[ j ] > R[ pos ] + road[ pos ][ j ]) {

                   R[ j ] = R[ pos ] + road[ pos ][ j ]; 

                   F[ j ] = pos;
                }
         }
     } 
}

void solution( node ) {
     
     int i;

     get(node);

     printf("The cost from 1 to %d = %d\nThe shortest path is -> ", ender, R[ender]);
 
     for(i=0;i<n;i++) {

         if(output[i]) printf("%d ", output[i]);
     }  
}

void get( node ) {

     if(F[ node ]) get( F[ node ] );

     output[ind++] = node;
}

void writeToFile(const char *filename) {

     int i;

     fout = fopen(filename,"w");

     fprintf(fout, "The cost from 1 to %d is %d\n The shortest path is -> \n", ender, R[ender]);
 
     for(i=0;i<n;i++) {

         if(output[i]) fprintf(fout, "%d ", output[i]);
     }
    
     fclose( fout );

}

File: dijkstra.py
# Dijkstra's algorithm conceived by computer scientist Edsger Dijkstra in 1956 and published in 1959
# is a graph search algorithm that solves the single-source shortest path problem for a graph with 
# non-negative edge path costs, producing a shortest path tree.
# O(N^3) Running Time complexity

class Dij: 

      #matrix of roads
      road = [[]]        

      #infinity
      infinit = 9999 

      #array of costs
      D = []

      #array of selected nodes
      S = []

      #array of fathers
      T = []
   
      #number of nodes  
      nodes = -1

      #output
      output = []

      def __init__(self, start):

          #read the graph from FILE.in  
          self.readGraph()

          #start node
          self.r = start

          #assign again
          r = self.r

          #step 1 you must to mark the selected start node in the utility vector called S 
          self.S[ r ] = 1    

          #step 2
          for j in range(1, self.nodes + 1):

              self.D[ j ] = self.road[r][j]

          #step 3
          for j in range(1, self.nodes+1):

              if self.D[ j ]:

                 self.T[ j ] = r 

          for i in range(1, self.nodes):

              #find the nodes that is the close to started node
              min = 9999

              #at nest step you must walk through array and find out which one is the minimum cost from start node.
              for j in range(1, self.nodes + 1):

                #if not selected
                if self.S[ j ] == 0:

                  if self.D[j] < min:

                     min = self.D[j]
                     pos = j

              self.S[ pos ] = 1
         
              #next step
              for j in range(1, self.nodes + 1):

                  if self.S[ j ] == 0:
                     if self.D[ j ] > self.D[ pos ] + self.road[ pos ][ j ]:
                        self.D[ j ] = self.D[ pos ] + self.road[ pos ][ j ]
                        self.T[ j ] = pos
     
          f = open('road.out','w')
           
          for k in range(1, self.nodes + 1):
              if k is not r:
                 if self.T[ k ]:

                    print "Road from ", r, " to ", k

                    self.draw(k)

                    print self.output

                    f.write("Road from {} to {} => {}".format(r,k,str(self.output)))

                    f.write("\n")

                    self.output = []

                 else:
                    print "There is not exists road"
                                
          f.close() 

      def draw(self, node):

          if self.T [ node ]:

             self.draw( self.T[ node ] )

          print node
          self.output.append(node)

      def readGraph(self):

          counter = 0

          input = []
  
          with open('graph.in', 'r') as file:

               for a_line in file:

                   counter += 1 
                   
                   if counter == 1:

                      number_of_nodes = int(a_line.rstrip())

                   else:

                      input.append(a_line.rstrip()) 

          size = len( input )

          self.nodes = number_of_nodes

          self.road = [[0 for i in range(0, number_of_nodes + 1)] for j in range(0, number_of_nodes + 1)]

          for i in range(0, self.nodes + 1):

              for j in range(0, self.nodes + 1):

                  if i == j:

                     self.road[i][j] = 0

                  else:

                     self.road[i][j] = self.infinit    

          for i in range(0, size):

              component = input[i]
              node1 = int(component[0])
              node2 = int(component[2])
              cost = int(component[4])  

              self.road[node1][node2] = cost

          #init
          self.D = [ 0 ]* (number_of_nodes + 1)
          self.S = [ 0 ]* (number_of_nodes + 1)
          self.T = [ 0 ]* (number_of_nodes + 1)
 


ob = Dij(2)
